### ISSUE TITLE
Map editor popup (tutorial / start editing) pops up every time

### ISSUE THREAD

**Issue by jancborchardt** [openstreetmap/iD#3002]
Currently every time I want to edit on OSM.org, this modal asking me if I’m interested in a tutorial (or just want to start editing) pops up every time. Yes, the new editor is awesome and no, I don’t need a tutorial. ;) At some point in the past I’m pretty sure I did the tutorial and it probably helped me. But now the popup is annoying – every time.

TL;DR: Only show this popup _once_ for a user. When I click »Edit« in the top bar, I want to directly edit. (I assume the tutorial can be accessed otherwise anyway if it’s needed later.)

As @tomhughes said in https://github.com/openstreetmap/openstreetmap-website/issues/1162#issuecomment-189285230, the mechanism for remembering if to show the popup or not is briwser local storage. It would be more appropriate to permanently store it with the user profile because otherwise it will pop up again when switching browsers or devices, clearing profiles, reinstalling etc – everything which has nothing to do with the editor.

**Comment by bhousel** [comment]
The store will definitely need to stay as the browser's `localStorage`.  There is no capability to store editor preferences in the user's OSM profile.

Have you disabled `localStorage`, or are running in incognito mode?

**Comment by bhousel** [comment]
BTW, I was wrong before and there actually is a way to store editor preferences with the user's OSM profile.
http://wiki.openstreetmap.org/wiki/API_v0.6#Preferences_of_the_logged-in_user

**Comment by DarkPrince304** [comment]
Hello @bhousel , @jancborchardt 

I would like to work on a fix for this. Has anyone already started looking into this? Or has this been fixed?

What I plan to do this aunthenticate with the logged in user and send a PUT request with a key and a value which will enable the user to remember his/her choice.

**Comment by bhousel** [comment]
I'm not sure whether we want to do this.  My initial feeling was to just close it as WONTFIX, but I decided to leave it open for discussion.  Do we really want to store user preferences in OSM?  

We do store a bunch of things in localStorage that could just as well be stored in OSM, but we'd need to add the necessary methods to `iD.Connection` to PUT/GET to these endpoints and serialize/deserialize the XML.  It just seems like a lot of work for very little benefit.

For this to work the user needs to be already authenticated.  This is the case when iD is embedded in OSM.org, but not necessarily when iD is used on other sites.

**Comment by pnorman** [comment]
There's not much precedent for what to store in the user preferences. Anything that is stored should be valid across multiple computers and editor instances, which the iD walkthrough remembering is. If you've gone through the iD walkthrough on osm.org, you don't need to on another site.

**Comment by jfirebaugh** [comment]
I'm in favor of storing preferences using the API. Maybe just store them all in one JSON blob?

**Comment by pnorman** [comment]
Since it stores key=values, I think it makes more sense to store individual values for most items.

**Comment by bhousel** [comment]
@jfirebaugh how about history too?  I have been regularly pushing my edits up into the many hundreds and overflowing the localStorage limit.

**Comment by bhousel** [comment]
Yes I'd also prefer separate `key=values` for flexibility in changing things.

**Comment by pnorman** [comment]
History isn't a preference, and storing it on the server means that you'll loose if you can't upload, which is when you need it the most.

**Comment by DarkPrince304** [comment]
Hello,

Even I think storing on the API Side would be a nice idea. I would like to look into this. So would need some guidance on this. I guess this will require a new pair of key-value? So how do I add this pair to the user preference list?

By the way on this site : http://www.openstreetmap.org/edit?editor=id that popup is visible but it isnt visible on www.openstreetmap.org. Is that the expected behaviour?

**Comment by bhousel** [comment]
@DarkPrince304 This is not a very good task for a new contributor, but if you really are interested my advice is to look in [`iD.Connection`](https://github.com/openstreetmap/iD/blob/master/js/id/core/connection.js) at how iD communicates with the OSM API.

There is an [`authenticated()`](https://github.com/openstreetmap/iD/blob/88670a9ccf0da18dd3119292984320dc7cf6d664/js/id/core/connection.js#L201) method that would tell you if the user is already logged in.  

[Code in the global `iD` context object](https://github.com/openstreetmap/iD/blob/88670a9ccf0da18dd3119292984320dc7cf6d664/js/id/id.js#L10-L32) manages localStorage, so that is the place where you'd need to check whether the user is authenticated and call the (not yet written) code in `iD.connection` to pull values from server side storage.  It would need to fall back to localStorage if the user is not authenticated, and ideally push local storage to server storage once authentication happens.

**Comment by tbertels** [comment]
See https://github.com/tyrasd/overpass-turbo/commit/d1933261c9d8bebb1f96b676871736454479a3d7 for an example of implementation of http://wiki.openstreetmap.org/wiki/API_v0.6#Preferences_of_the_logged-in_user.

**Comment by bhousel** [comment]
This relates to #3239 / #4418 and would land as an iD v3 feature..

There are currently assumptions in the code that a user's settings are available synchronously at startup (e.g. to show the splash screen and restore history), and introducing an asynchronous dependency would be a breaking change.

**Comment by jidanni** [comment]
I was still able to trigger the "Welcome to the iD OpenStreetMap editor" by directly clicking "edit" from the osm.org homepage when not logged in.
I think the assumption is the user will first click "Log in", then after logging in, then click "Edit".
However if he instead first clicks "Edit", (then fills in his password,) he will trigger the "Welcome to the iD OpenStreetMap editor" pop-up (despite having used iD lots before.)

**Comment by tbertels** [comment]
@jidanni: I actually got it will being logged in and clicking on Edit, so I think it's because of the new privacy policy (#7040).

**Comment by jtracey** [comment]
The preference to show third party icons, which the UI implies will be stored, is currently not. I'm assuming fixing that would be part of this bug?

**Comment by quincylvania** [comment]
@jtracey Third-party icons aren't intended to be stored between sessions. The privacy option is about whether to load them at all.

**Comment by jtracey** [comment]
@quincylvania I know, I'm saying the setting to not do that isn't stored (i.e., I have to disable it every time I load iD).

**Comment by bhousel** [comment]
> I'm saying the setting to not do that isn't stored (i.e., I have to disable it every time I load iD).

The preference should be stored, unless you are using something like incognito mode, or maybe a privacy-focused web browser that throws out the storage.

**Comment by jtracey** [comment]
I delete local storage when I close the browser, yes. Which is exactly what this issue is about ameliorating, no?

**Comment by bhousel** [comment]
> I delete local storage when I close the browser, yes. Which is exactly what this issue is about ameliorating, no?

Yes.
